src/mlmc/sampler.py

```diff
@@ -0,0 +1,181 @@
+import os
+import time
+import numpy as np
+from typing import List
+
+from sample_storage import SampleStorage
+from sampling_pool import SamplingPool
+from new_simulation import Simulation
+from workspace import WithoutWorkspace, SimulationWorkspace, WholeWorkspace, PBSWorkspace
+
+
+class Sampler:
+
+    def __init__(self, sample_storage: SampleStorage, sampling_pool: SamplingPool, sim_factory: Simulation,
+                 n_levels: int, step_range: List[int], work_dir=None):
+        """
+
+        :param sample_storage: store scheduled samples, results and result structure
+        :param sampling_pool: calculate samples
+        :param sim_factory: generate samples
+        :param n_levels: number of MLMC levels
+        :param step_range: simulation step range
+        :param work_dir: working directory
+        """
+        self._sample_storage = sample_storage
+        self._sampling_pool = sampling_pool
+
+        self._n_levels = n_levels
+        self._sim_factory = sim_factory
+        self._step_range = step_range
+
+        # Number of created samples
+        self._n_created_samples = np.zeros(self._n_levels)
+
+        # Number of target samples
+        self.n_target_samples = np.zeros(self._n_levels)
+        # Number of planned samples
+        self.n_planned_samples = np.zeros(self._n_levels)
+        # Number of finished and collected samples
+        self.n_collected_samples = np.zeros(self._n_levels)
+
+        # Set workspaces
+        self._create_workspace_directories(work_dir)
+
+        # Save simulation result specification
+        self._sample_storage.save_workspace(self._workspace)
+
+    def _create_workspace_directories(self, work_dir):
+        """
+        Sampler uses workspace object for communication with files system
+        :param work_dir: abs path to working directory
+        :return: None
+        """
+        try:
+            self._workspace = WithoutWorkspace()
+
+            if work_dir:
+                self._workspace = WholeWorkspace(work_dir, self._n_levels)
+                try:
+                    if self._sampling_pool.need_jobs:
+                        self._workspace = PBSWorkspace(work_dir, self._n_levels)
+                        self._sampling_pool.workspace = self._workspace
+                except AttributeError:
+                    print("PBS sampling pool must have need_jobs attribute")
+
+            # if self._sim_factory.need_workspace:
+            #     self._workspace = SimulationWorkspace(work_dir)
+        except AttributeError:
+            print("Attribute error, need_workspace class attribute is obligatory")
+
+    def determine_level_n_samples(self):
```

Comment:
There should be a method to update target n_samples for given variance estimates
and given target variance using actual sample times from collected samples.